blocksync/pool.go

```diff
@@ -633,9 +632,26 @@ OUTER_LOOP:
 				}
 			case <-bpr.gotBlockCh:
 				// We got a block!
-				// Continue the for-loop and wait til Quit.
-				continue WAIT_LOOP
+				// Exit the for loop and wait until we receive a quit signal.
+				break WAIT_LOOP
+			case <-bpr.pool.Quit():
+				if err := bpr.Stop(); err != nil {
+					bpr.Logger.Error("Error stopped requester", "err", err)
+				}
+				return
+			case <-bpr.Quit():
+				return
+			}
+		}
+
+		select {
+		case <-bpr.pool.Quit():
+			if err := bpr.Stop(); err != nil {
+				bpr.Logger.Error("Error stopped requester", "err", err)
 			}
+			return
+		case <-bpr.Quit():
+			return
```

Comment:
Factoring this out is not really necessary for this change. However, the intermingling here seemed strange: once the block was received, the for loop would be restarted so that the requester could wait to be shutdown by receiving the quit signal. However, this also allowed the successful requester to receive a 'redo' signal, prompting it to run the block request again even though it would have already received the block. This change moves the shutdown logic into this separate select clause that is run after the block is received to prevent this odd case.

Reply:
Note that redo can be triggered if validation fails on a received block. The intent there is to redo the request, by removing the peer which you requested the block from and doing the `continue OUTER_LOOP` thing within this case statement.  But I think your logic is still correct in that we should make sure redo is not ran for a block that is validated and applied.

Reply:
From @jmalicevic's comment, I think the changes here could be minimized by not touching other conditions, and the newly added condition (timeout) finishing with `continue OUTER_LOOP`... effectively turning the timeout case, into a block validation failure.

HOWEVER, as I am taking over from William, I will merge this as is (@jmalicevic says the logic is still correct); will backport to `v0.37.x`; and run the _rotating nodes_ test with this change in.

Then, when William's back, we can rediscuss if we want to simplify it in `main`.